### PR TITLE
Add logic to automatically enable consensus system

### LINKED ISSUE #571: Update ensemble bootstrap logic to enable consensus
In basho/riak_ensemble#10, the clustering system for `riak_ensemble` was changed to require a manual step to enable the consensus system on a single node before joining nodes together into a cluster. This change was made to ensure that there is only ever a single logical root ensemble history. Joining nodes adopt the history from the single enabled node.

Thus, a user must now run `riak_ensemble_manager:enable().` from the console of Riak claimant to enable strong consistency.

We should change the bootstrap logic in `riak_core_claimant` to perform this step automatically. The primary challenge is ensuring only a single node ever performs that step, even in the presence of node failures/partitions.

One way to implement this is to have the current claimant wait for "ring ready" and then check for a ring metadata key that notes the consensus singleton node. If the key is missing, the claimant writes itself as the singleton and waits yet again for "ring ready". If the claimant sees itself as the singleton in a ready ring, the claimant then enables the consensus system. Likewise, the bootstrap logic would be updated to join nodes to the known singleton (based on the metadata value) rather than the current claimant.

/cc basho/riak#536

### PR DESCRIPTION
**Pull request by jtuple**
Currently, in addition to enabling consensus in app.config, a user must also manually call `riak_ensemble_manager:enable()` from one and only one node in a cluster to activate the consensus sub-system. This is necessary to ensure that there is only a single logical root ensemble history -- all other nodes adopt the history from the single enabled node.

However, this step is not only annoying but also error-prone. Enabling consensus on multiple nodes can break the consensus system, requiring manual intervention.

This commit addresses this problem by making riak_core automatically enable the consensus system in a safe way. This is accomplished by having the claimant node enable the consensus system. To avoid the issue where the claimant in multiple 1-node clusters enables consensus before being joined, this commit requires the cluster to have at least three nodes before the claimant will enable the consensus system.

To prevent a race during claimant changes, a claimant must first write a special ring metadata value that prevents future claimants from activating the consensus system. It is not until after the ring has converged cluster wide, and the claimant sees the appropriate metadata value, that the claimant activates the consensus system.

Resolves #571

Sibling pull-request: basho/riak_test#632

For testing purposes, the existing `riak_ensemble` tests from `riak_test` work well -- just make sure to use the sibling `riak_test` pull-request above.

### CODE REVIEW

**Comment by evanmcc** [comment]
with the fixes: 

Test Results:
ensemble_remove_node2-riak_kv_bitcask_backend,riak_kv_memory_backend: fail
ensemble_remove_node-riak_kv_bitcask_backend,riak_kv_memory_backend : fail
ensemble_interleave-riak_kv_bitcask_backend,riak_kv_memory_backend  : fail
ensemble_basic4-riak_kv_bitcask_backend,riak_kv_memory_backend      : pass
ensemble_basic3-riak_kv_bitcask_backend,riak_kv_memory_backend      : pass
ensemble_basic2-riak_kv_bitcask_backend,riak_kv_memory_backend      : pass
ensemble_basic-riak_kv_bitcask_backend,riak_kv_memory_backend       : pass

without: 

Test Results:
ensemble_remove_node2-riak_kv_bitcask_backend,riak_kv_memory_backend: fail
ensemble_remove_node-riak_kv_bitcask_backend,riak_kv_memory_backend : fail
ensemble_interleave-riak_kv_bitcask_backend,riak_kv_memory_backend  : fail
ensemble_basic4-riak_kv_bitcask_backend,riak_kv_memory_backend      : fail
ensemble_basic3-riak_kv_bitcask_backend,riak_kv_memory_backend      : fail
ensemble_basic2-riak_kv_bitcask_backend,riak_kv_memory_backend      : fail
ensemble_basic-riak_kv_bitcask_backend,riak_kv_memory_backend       : fail

**Comment by evanmcc** [comment]
failures on the patched code are 2x `node_shutdown_timed_out` and one:

```
17:47:26.579 [warning] ensemble_interleave failed: {{assertEqual_failed,[{module,rtdev},{line,260},{expression,"file : write_file ( ConfigFile , NewConfigOut )"},{expected,ok},{value,{error,enoent}}]},[{rtdev,'-update_app_config_file/2-fun-3-',3,[{file,"src/rtdev.erl"},{line,260}]},{rtdev,update_app_config_file,2,[{file,"src/rtdev.erl"},{line,260}]},{rt,'-pmap/2-fun-0-',4,[{file,"src/rt.erl"},{line,1497}]}]}
17:47:26.579 [error] 
================ ensemble_interleave failure stack trace =====================
{{assertEqual_failed,[{module,rtdev},
                      {line,260},
                      {expression,"file : write_file ( ConfigFile , NewConfigOut )"},
                      {expected,ok},
                      {value,{error,enoent}}]},
 [{rtdev,'-update_app_config_file/2-fun-3-',3,
         [{file,"src/rtdev.erl"},{line,260}]},
  {rtdev,update_app_config_file,2,[{file,"src/rtdev.erl"},{line,260}]},
  {rt,'-pmap/2-fun-0-',4,[{file,"src/rt.erl"},{line,1497}]}]}
==============================================================================
```

**Comment by evanmcc** [comment]
ignore the two above comments, riak_ensemble was out of date, rerunning.

File: src/riak_core_claimant.erl [hunk 1]
```diff
@@ -746,6 +747,90 @@ type_claimant(Props) ->
...
     end.
 
+%% The consensus subsystem must be enabled by exactly one node in a cluster
+%% via a call to riak_ensemble_manager:enable(). We accomplished this by
+%% having the claimant be that one node. Likewise, we require that the cluster
+%% includes at least three nodes before we enable consensus. This prevents the
+%% claimant in a 1-node cluster from enabling consensus before being joined to
+%% another cluster.
+maybe_enable_ensembles() ->
+    Desired = riak_core_sup:ensembles_enabled(),
```
Comment (evanmcc) [inline src/riak_core_claimant.erl:757]: I assume that one of these calls has a capabilities check hidden in it?

**Comment by evanmcc** [comment]
+1 8d2cda5

with the patch:

```
ensemble_remove_node2-riak_kv_bitcask_backend,riak_kv_memory_backend: pass
ensemble_remove_node-riak_kv_bitcask_backend,riak_kv_memory_backend : pass
ensemble_basic4-riak_kv_bitcask_backend,riak_kv_memory_backend      : pass
ensemble_basic3-riak_kv_bitcask_backend,riak_kv_memory_backend      : pass
ensemble_basic2-riak_kv_bitcask_backend,riak_kv_memory_backend      : pass
ensemble_basic-riak_kv_bitcask_backend,riak_kv_memory_backend       : pass
```

without: 

```
ensemble_remove_node2-riak_kv_bitcask_backend,riak_kv_memory_backend: fail
ensemble_remove_node-riak_kv_bitcask_backend,riak_kv_memory_backend : fail
ensemble_basic4-riak_kv_bitcask_backend,riak_kv_memory_backend      : fail
ensemble_basic3-riak_kv_bitcask_backend,riak_kv_memory_backend      : fail
ensemble_basic2-riak_kv_bitcask_backend,riak_kv_memory_backend      : fail
ensemble_basic-riak_kv_bitcask_backend,riak_kv_memory_backend       : fail
```

code looks good, tests pass.

File: src/riak_core_claimant.erl (hunk 1 above)
Comment (jtuple) [inline src/riak_core_claimant.erl:757]: Sorta. `riak_core_sup:ensembles_enabled()` has an `app.config` check, and we currently ship w/ consensus subsystem disabled. Assumption was folks would manually change the setting after upgrading all nodes. But, I was just thinking about a capability the other day. We probably should add one before 2.0 final. But, that'll be in a later pull-request.

**Comment by jtuple** [comment]
@borshop merge

**Pull request merged**